source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveActionRenderer.cs

```diff
@@ -16,40 +16,7 @@ public static FrameworkElement Render(AdaptiveAction action, AdaptiveRenderConte
                 {
                     if (action is AdaptiveToggleVisibilityAction toggleVisibilityAction)
                     {
-                        foreach (object targetElement in toggleVisibilityAction.TargetElements)
-                        {
-                            string targetElementId = "";
-                            bool? targetElementIsVisible = null;
-
-                            if (targetElement is string targetElementString)
-                            {
-                                targetElementId = targetElementString;
-                            }
-                            else if (targetElement is AdaptiveTargetElement targetElementObject)
-                            {
-                                targetElementId = targetElementObject.ElementId;
-                                targetElementIsVisible = targetElementObject.IsVisible;
-                            }
-
-                            var element = LogicalTreeHelper.FindLogicalNode(context.CardRoot, targetElementId);
-
-                            if (element != null && element is FrameworkElement elementFrameworkElement)
-                            {
-                                Visibility visibility = elementFrameworkElement.Visibility;
-                                // if we read something with the format {"elementId": <id>", "isVisible": true} or we just read the id and the element is not visible
-                                if ((targetElementIsVisible.HasValue && targetElementIsVisible.Value) || (!targetElementIsVisible.HasValue && visibility != Visibility.Visible))
-                                {
-                                    elementFrameworkElement.Visibility = Visibility.Visible;
-                                }
-                                // otherwise if we read something with the format {"elementId": <id>", "isVisible": false} or we just read the id and the element is visible
-                                else if ((targetElementIsVisible.HasValue && !targetElementIsVisible.Value) || (!targetElementIsVisible.HasValue && visibility == Visibility.Visible))
-                                {
-                                    elementFrameworkElement.Visibility = Visibility.Collapsed;
-                                }
-
-                            }
-
-                        }
+                        context.ToggleVisibility(toggleVisibilityAction.TargetElements);
```

Comment:
>ToggleVisibility [](start = 32, length = 16)

For what it's worth, in UWP i just put this under the equivalent of InvokeAction rather than having a separate function. That is, if you call InvokeAction on a toggle action it toggles the relevant elements. What do you think of that approach vs. this one?

Reply:
One use case I was thinking of is if a host creates their own `MySuperAction` which lets you (1) toggle visibilities, (2) submit something via HTTP, and (3) toggle visibilities again after submission completed, it ought to be easy for the host to perform those. However, with that approach, it would require the host creating a temporary Action.ToggleVisibility and then calling InvokeAction on it?

Reply:
Fair enough, but if the host wants to trigger actions themselves, for example they've got custom action renderers, in the approach where you use a single function the host doesn't have to have this if (they can just call InvokeAction on all actions). BTW, does show card work with InvokeAction in dotnet? 

---
In reply to: [268849214](https://github.com/Microsoft/AdaptiveCards/pull/2559#discussion_r268849214) [](ancestors = 268849214)

Reply:
I just discovered WPF doesn't support inline ShowCard :O it only supports popup.

But agreed, we should lump this logic into InvokeAction so hosts can just call that on all actions. I think in addition to that, we should still keep `ToggleVisibility` exposed for more custom use cases. I'll take a look at the UWP code and see how that looks and update the pull request.

Reply:
Alright, I updated it so that all a host does is call `InvokeAction` and pass the action!

```csharp
                uiButton.Click += (sender, e) =>
                {
                    context.InvokeAction(uiButton, new AdaptiveActionEventArgs(action));
```

The renderer code then determines if it's a ToggleVisibility and handles it, otherwise it sends out the event to the host

```csharp
        public void InvokeAction(FrameworkElement ui, AdaptiveActionEventArgs args)
        {
            // ToggleVisibility is a renderer-handled action
            if (args.Action is AdaptiveToggleVisibilityAction toggleVisibilityAction)
            {
                this.ToggleVisibility(toggleVisibilityAction.TargetElements);
                return;
            }

            OnAction?.Invoke(ui, args);
        }
```

I think we should still keep `ToggleVisibility` public on the Context too, so that hosts could use that like I was describing. For example, hosts could extend validation and do something like this...

```json
{
  "type": "Input.Text",
  "validation": {
    "whenValidToggle": [
      { "id": "successText", "isVisible": true }
    ]    
```

And then they just pass that list to `context.ToggleVisibility`

Reply:
Should we log a bug for inline show cards for WPF?

---
In reply to: [268867015](https://github.com/Microsoft/AdaptiveCards/pull/2559#discussion_r268867015) [](ancestors = 268867015)

Reply:
👍 Filed issue #2599